### PR TITLE
Add a ProcessMonitor to help receive more consistent observation for the monitor

### PR DESCRIPTION
**Pull request by Delaunay**

### CODE REVIEW

File: voir/instruments/utils.py [hunk 1]
```diff
@@ -20,3 +21,92 @@ def run(self):
...
+
+    def stop(self):
+        for a in self.args:
+            a.stop()
+
+
+def monitor(delay, getfun, pushfun, process=True):
+    """Run the monitor in a different process to have metrics in regular intervals
+    Pusher is a thread that gets executed when there is time.
+    """
```
Comment (breuleux) [inline voir/instruments/utils.py]: Documentation for the arguments would be helpful.

File: voir/instruments/utils.py [hunk 2]
```diff
@@ -20,3 +21,92 @@ def run(self):
...
+    if process:
+        m = []
+
+        # Note: monitor needs to be first, so it stops generating observation first
+        monitor = ProcessMonitor(delay, getfun)
+        m.append(monitor)
+        if pushfun is not None:
```
Comment (breuleux) [inline voir/instruments/utils.py]: In what situation might `pushfun` be None? Also, if `pushfun` can be None, I reckon that should be its default value in the function's arguments.

File: voir/instruments/utils.py [hunk 3]
```diff
@@ -20,3 +21,92 @@ def run(self):
...
+            m.append(ProcessPusher(delay, monitor.results, pushfun))
+
+        return _Monitor(*m)
+
+    #
+    def fun():
+        pushfun(getfun())
```
Comment (breuleux) [inline voir/instruments/utils.py:119]: If `pushfun` is None, this will crash.

File: voir/instruments/utils.py [hunk 4]
```diff
@@ -20,3 +21,92 @@ def run(self):
...
+
+        # Note: monitor needs to be first, so it stops generating observation first
+        monitor = ProcessMonitor(delay, getfun)
+        m.append(monitor)
+        if pushfun is not None:
+            m.append(ProcessPusher(delay, monitor.results, pushfun))
+
+        return _Monitor(*m)
+
+    #
```
Comment (breuleux) [inline voir/instruments/utils.py]:
```suggestion
```

File: voir/instruments/utils.py (hunk 1 above)
Comment (Delaunay) [reply voir/instruments/utils.py]: added

File: voir/instruments/utils.py (hunk 2 above)
Comment (Delaunay) [reply voir/instruments/utils.py]: You are right I think it is better to remove, IIRC it was for the case when the regular thread monitor was used.

File: voir/instruments/utils.py (hunk 3 above)
Comment (Delaunay) [reply voir/instruments/utils.py:119]: `pushfun` should not be none now

**Review by breuleux — APPROVED** [review]

**Pull request merged**